### ISSUE TITLE
move v0.3 to ES6

### ISSUE THREAD

**Issue by ssomnoremac** [futurepress/epub.js#434]
I'm not the first to mention it, and we could arguably attract more contributors interest if we converted the project to ES6 javascript.

Pros:
- initial conversion to ES6 classes and `let` and `const` is [automatic](http://lebab.io/)
- i've already done [that part](https://github.com/libraryforall/epub.js/tree/es6)
- replaces all `class.prototype.method = function()` definitions with `method()`
- maybe replace RSVP with something else RxJS? (or just stop using deferred)
- replace references to core.extend, etc. with spread operators and other ES6 native tools
- overall readability and comprehension improved
- ES6 is more fun!

Cons:
- would need to add babel and probably webpack

**Comment by JakeHartnell** [comment]
Rollup works pretty well: https://github.com/rollup/rollup
Example project: https://github.com/rollup/rollup-starter-project

**Comment by ssomnoremac** [comment]
I tried to bundle [my fork](https://github.com/libraryforall/epub.js/tree/es6) with webpack but I'm new webpack and the bundle is broken. Does anyone what the [webpack config](https://github.com/libraryforall/epub.js/blob/es6/webpack.config.js) should be based on what is currently in the `gulpfile.js` ? Or maybe just keep gulp? seems silly to have babel, gulp, and webpack though.

**Comment by ssomnoremac** [comment]
What's really great about this is that getters like this:

    Contents.prototype.scrollWidth = function() {
      var width = this.documentElement.scrollWidth;
      return width;
    };

can be replaced by one liners like this:

    scrollWidth = () => this.documentElement.scrollWidth

**Comment by fchasen** [comment]
I switched v0.3 over to using webpack for generating the bundle.

Will be easy to add babel in now, but still not sure it makes sense to switch all the code over to es6 at this point in time.

**Comment by gerhardsletten** [comment]
@ssomnoremac I have added a pr #503 that showcase an example with webpack

Also strongly vote for moving to ES6. Typically you then would only store the es6 version in git, add a prepublish command that compile it to es5 and then point the npm-published version to that. No need to store the build version in the git-library, people can install it through npm.

[Example of the prepublish command can be seen here.](https://github.com/gerhardsletten/react-reader/blob/master/package.json#L8)

**Comment by fchasen** [comment]
That sounds good (and should probably move the current webpack build to use prepublish) 

Moving to es6 will happen eventually, I was waiting for Node to support imports so it hasn't been a big priority to me, but maybe it should be done before we release v0.3.

Also reading up on RollUp a bit has made me think it seems more appropriate for bundling the distribution a library then webpack.

**Comment by gerhardsletten** [comment]
The discussion on ES6 is not so much about what's supported in node/browser x, but rather having the the best tools available when writing the code and code in its purest form. 

When the source is transpiled down to es5, you dont want to bundle, but transpile each file. Then in the .gitignore you would [ignore the es5 files](https://github.com/gerhardsletten/react-reader/blob/master/.gitignore#L36) and only store the source, and in the .npmignore you would do the opposite, [ignore the source](https://github.com/gerhardsletten/react-reader/blob/master/.npmignore#L1) and only include the es5 transpiled version.

With and index in the source-folder the user would be able to also be able to import only one module, [see an example of it there](https://github.com/gerhardsletten/react-reader/blob/master/src/index.js). 

This way you will only commit changes to the source, and the npm module contains es5 files.

The "externals" should only be defined as npm dependencies, and you should use a browser-ready path replacement like this instead like [path-browser-wrapper](https://github.com/jbhoosreddy/path-browser-wrapper)

**Comment by fchasen** [comment]
I like the idea of being able to `require` directly with node (with no transpile needed) but the benefit of cleaning up the code will likely outweigh that.

I'll give switching to es6 imports and classes an initial go this weekend and push a branch up to continue this discussion.

That path wrapper (and others) are missing `resolve` and `relative` which I've added to the one v0.3 is using: https://github.com/fchasen/path-webpack

**Comment by gerhardsletten** [comment]
Great! 

> I like the idea of being able to require directly with node (with no transpile needed) but the benefit of cleaning up the code will likely outweigh that.

If you use nvm ist easy use multiple versions of node, and then you can use the latest which should support almost everything in es6

> That path wrapper (and others) are missing resolve and relative which I've added to the one v0.3 is using: https://github.com/fchasen/path-webpack

Are epub.js supposed to work in a server-enviroment as well? Or is it only to resolve the urls within an unzipped epub file? If it is the latter you could just import the module this way:

```js
import path from 'path-webpack'
```

And then you should not need to define it as an alias in webpack.config (and all other who use epub.js would not as well ;-)

**Comment by fchasen** [comment]
PR with the initial switch to ES2015 is here: https://github.com/futurepress/epub.js/pull/516

Very happy with the babel `env` preset letting us target specific browser releases to support for the `dist` version. I've set the browser requirement very high for `epub.js` and added a `epub.legacy.min.js` with the polyfills included to support older browsers.

I've switched to just including path as suggested above as well.
